### PR TITLE
Transport-level query batching

### PR DESCRIPTION
**Pull request by Poincare**
Adds transport-level query batching support to Apollo Client (see #505).

TODO:

- [x] Update CHANGELOG.md with your change
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

### CODE REVIEW

**Comment by stubailo** [comment]
This is great. We should do some things before merging and shipping the feature:

1. Which server implementations currently support batching? Of course, apollo-server. What about graphql-ruby, sangria, express-graphql with https://github.com/nodkz/react-relay-network-layer? Can we make this solution work with those as well? There is no spec for how this stuff should work, but perhaps we can build something close enough that people can use it with different servers. (see here as well: https://github.com/graphql/express-graphql/pull/100)
2. We should write documentation, both for how to use this feature on the client and how one might add this to their server (with appropriate links to various server docs etc)

**Comment by nodkz** [comment]
> @nodkz I'm curious - what is the id for in your react-relay-network-layer? I think this is the right API and @Poincare is in the process of implementing batching for Apollo Client - would it be reasonable for us to just implement an array of requests and array of responses?

**Shortly, `id` is the Relay request id.**

When Relay provides a list of queries for fetching. I can pass it back in any order, but should know for which query which payload addressed. So I combine all queries with id into an array, then send them to the server. Server proceeds all queries and passes array payloads with its ids back. Client gets the response, traversing by the array and maps by `id` which payload corresponds to which relay query.  
https://github.com/nodkz/react-relay-network-layer/blob/master/src/relay/queriesBatch.js#L6 

Example of batch request:
```js
[
  {
    "id": "q0",
    "query": `
      query ViewerQueries {
        viewer {
          ...F0
        }
      }
      fragment F0 on Viewer {
        id,
        usersCount
      }
    `,
    "variables": {}
  },
  {
    "id": "q1",
    "query": "query ViewerQueries { ... }",
    "variables": {}
  }
]
```

Batch response:
```js 
  { 
    "id":"q0", 
    "payload":{
      "data": {
        "viewer": {
          "id": "Q2FiaW5ldDo1N2FjOWE2NDVhZTUxYTI1ZmYxZGNkMzg=",
          "usersCount": 0,
        }
      }
    } 
  },
  { 
    "id":"q1", 
    "payload": {
      "data": null,
      "errors": [ ... ]
    } 
  }
]
```

**Comment by stubailo** [comment]
OK, so the idea is that Relay batching is built with the idea that requests might arrive out of order or separately, even if they were sent as a batch? This seems like it would never come up with HTTP batching, right? So maybe we could remove the `id` from the HTTP batching implementation?

**Comment by nodkz** [comment]
@stubailo if https://github.com/graphql/express-graphql/issues/100 will be accepted, then I'll remove `id` from request, cause `express-graphql` will guarantee order of returned payloads according to queries.

But if I realize network layer via WebSockets, then I'll return `ids` back. To have ability pass fetched data to relay, while other queries in progress.

**Comment by Poincare** [comment]
I'm not sure if batching is necessary at all over a WebSocket transport since the concept of a "roundtrip" no longer exists.

**Comment by nodkz** [comment]
Nice catch. @Poincare totally agreed with you.

**Comment by nodkz** [comment]
Ooops. Not agree again.

Suppose the following scenario: We have 4 queries from relay `q1, q2, q3, q4`, and put them all to WebSocket.
We should wrap it in something with `id` that will be correlated returned payloads in different order, e.g. `p3, p1, p4, p2`.

File: src/networkInterface.ts [hunk 1]
```diff
@@ -197,35 +204,45 @@ export function createNetworkInterface(uri: string, opts: RequestInit = {}): HTT
...
+// https://github.com/Microsoft/TypeScript/issues/21
+import {
+  HTTPBatchedNetworkInterface,
+} from './batchedNetworkInterface';
+
+
+export function createNetworkInterface(
+  uri: string,
+  opts: RequestInit = {},
+  transportBatching = false
```
Comment (stubailo) [inline src/networkInterface.ts]: I don't think just having this as a third argument is a good idea. I'd replace `uri` with `NetworkInterfaceOptions` or something.

So you can do `createNetworkInterface('/graphql')` or `createNetworkInterface({ uri: '/graphql', serverBatching: true })`

Comment (Poincare) [inline src/networkInterface.ts]: Should `opts` be a part of `NetworkInterfaceOptions` or should it be a second argument? For example, would you do:

```
createNetworkInterface({
  uri: '/graphql',
  opts: { ... },
  serverBatching: true,
})
```

or

```
createNetworkInterface({
  uri: '/graphql',
  serverBatching: true,
},
{ ... })
```

The first option makes more sense to me but it probably breaks existing code that uses `createNetworkInterface`.

Comment (stubailo) [inline src/networkInterface.ts]: We can do the first one in a backwards-compatible way by checking if the first argument is a `string`. then for 1.0 we can remove the backcompat?

Comment (Poincare) [inline src/networkInterface.ts]: Sounds good to me.

File: test/client.ts [hunk 2]
```diff
@@ -1651,7 +1651,11 @@ describe('client', () => {
       },
       result: createMockedIResponse([firstResult, secondResult]),
     });
-    const networkInterface = createNetworkInterface('http://not-a-real-url.com', {}, true);
+    const networkInterface = createNetworkInterface({
+      uri: 'http://not-a-real-url.com',
+      opts: {},
```
Comment (stubailo) [inline test/client.ts]: What does `opts` do?

File: src/networkInterface.ts [hunk 3]
```diff
@@ -232,17 +232,39 @@ import {
...
+  if (isString(interfaceOpts) || !interfaceOpts) {
+    const uri = interfaceOpts as string;
+    return addQueryMerging(new HTTPFetchNetworkInterface(uri, backOpts)) as HTTPNetworkInterface;
   } else {
-    // createNetworkInterface has batching ability by default through query merging,
-    // which is not used unless the `shouldBatch` option is passed to apollo-client.
-    return addQueryMerging(new HTTPFetchNetworkInterface(uri, opts)) as HTTPNetworkInterface;
+    const {
+      transportBatching = false,
+      opts = {},
```
Comment (stubailo) [inline src/networkInterface.ts]: Can we change this to a more descriptive name since before it was only internal? Like `fetchOptions`.

**Comment by helfer** [comment]
@Poincare @stubailo what's left to do here other than rebasing? If possible, I want to merge this next week.

**Comment by Poincare** [comment]
This PR works (tested on GitHunt), but if we merge the typescript compiler options PR first, it will require a few changes so it is waiting for that.

**Pull request merged**